lib: smf: Fix smf_terminate sequence

TomzBench opened the pull request:
Currently, if the state machine is terminated, you have to run the state
machine one more time in order to bubble up the return code.  I think
it would be more precise to have the return code passed to the caller
during the same instance that the state machine was terminated in.  This
seems more sane, and also makes writing unit tests easier for smf
dependant code.

Most of the changes in this PR are to the unit tests, because they no
longer need to use a for loop for driving the state machine, since we
can now calculate when the state machine will terminate more clearly.

Signed-off-by: thomas <thomas.chiantia@gmail.com>

keith-zephyr commented on tests/lib/smf/src/test_lib_flat_smf.c:
```diff
@@ -36,8 +36,8 @@
 #define TEST_ENTRY_VALUE_NUM     0
-#define TEST_RUN_VALUE_NUM       4
-#define TEST_EXIT_VALUE_NUM      8
+#define TEST_RUN_VALUE_NUM       1
```
Note - SMF_RUN define at line 24 is no longer used.

TomzBench replied:
replaced with SMF_NUM_STATES, which is used in transition test

keith-zephyr commented on tests/lib/smf/src/test_lib_flat_smf.c:
```diff
@@ -204,7 +214,7 @@ static void state_d_entry(void *obj)
 static void state_d_run(void *obj)
 {
-	/* Do nothing */
+	smf_set_terminate(obj, 1);
```
I'd prefer a macro for the terminate value here - `STATE_D_TERMINATE_VALUE`.
It would also be an improvement to change the other termination calls to use unique terminate values.

keith-zephyr commented on tests/lib/smf/src/test_lib_flat_smf.c:
```diff
@@ -244,11 +258,8 @@ void test_smf_flat(void)
-		}
-	}
+	ret = smf_run_state((struct smf_ctx *)&test_obj);
+	zassert_equal(-1, ret, "Incorrect smf_run_state return value");
```
Nit - the comment at 255 confused me at first as I thought the "B)" label was referring to a state name.  Can you remove all the labels?  `/* A) Test transitions */` becomes `/* Test transitions */`

TomzBench replied:
fixed

keith-zephyr commented on tests/lib/smf/src/test_lib_flat_smf.c:
```diff
@@ -223,15 +233,19 @@ void test_smf_flat(void)
-			break;
-		}
-	}
+	ret = smf_run_state((struct smf_ctx *)&test_obj);
+	zassert_equal(0, ret, "Incorrect smf_run_state return value");
+	ret = smf_run_state((struct smf_ctx *)&test_obj);
+	zassert_equal(0, ret, "Incorrect smf_run_state return value");
+	ret = smf_run_state((struct smf_ctx *)&test_obj);
+	zassert_equal(0, ret, "Incorrect smf_run_state return value");
+	ret = smf_run_state((struct smf_ctx *)&test_obj);
+	zassert_equal(1, ret, "Incorrect smf_run_state return value");
```
Because the tests below check that the final state gets reach, I prefer calling `smf_run_state()` in a loop, as this better matches what application code will do.

TomzBench replied:
fixed using for loop.  Transition tests run SMF_NUM_STATES times, while each exit code test will run 1, 2, or 3 times, depending on the exit state (ENTRY, RUN, EXIT respectively)

keith-zephyr commented on tests/lib/smf/src/test_lib_flat_smf.c:
```diff
@@ -102,6 +102,11 @@ static void state_a_run(void *obj)
 	zassert_equal(o->transition_bits, test_value[o->tv_idx],
 		      "Test State A run failed");
 
+	if (o->terminate == RUN) {
+		smf_set_terminate(obj, -1);
+		return;
+	}
+
```
I like the original approach better, where each state supported different termination options.  Now you're always forcing state A to terminate for the entry/run/exit actions.

TomzBench replied:
fixed

sambhurst commented on tests/lib/smf/src/test_lib_flat_smf.c:
```diff
@@ -21,7 +21,11 @@
+#define STATE_C_TERMINATE_VALUE  -3
+#define STATE_D_TERMINATE_VALUE  -4
 
 #define STATE_A_ENTRY_BIT       (1 << 0)
```
nit - BIT(0)

TomzBench replied:
fixed

keith-zephyr reviewed (changes requested):
The changes look good.  However, can you squash the commits together?  As is, you're applying changes to the SMF tests in the first commit that you undo in the 2nd commit.  The overall changes are small enough that a single commit is fine.

TomzBench committed:
lib: smf: Fix smf_terminate sequence

Currently, if the state machine is terminated, you have to run the state
machine one more time in order to bubble up the return code.  I think
it would be more precise to have the return code passed to the caller
during the same instance that the state machine was terminated in.  This
seems more sane, and also makes writing unit tests easier for smf
dependant code.

Most of the changes in this PR are to the unit tests, because they no
longer need to use a for loop for driving the state machine, since we
can now calculate when the state machine will terminate more clearly.

Signed-off-by: thomas chiantia <thomas.chiantia@gmail.com>

TomzBench commented:
I'm not sure how to do that when the commits are already pushed.  I can squash these into a new branch and then i think a new PR will be created. Is this what you mean?

keith-zephyr commented:
> I'm not sure how to do that when the commits are already pushed. I can squash these into a new branch and then i think a new PR will be created. Is this what you mean?

If you squash the changes on your local branch, but then force push using `git push -f`, I think it will update the PR to match.

TomzBench commented:
ah, I didn't know -f would work.

It is all under the first commit now.

TomzBench commented:
@sambhurst  I think you approved this but was lost when added another fix for @keith-zephyr .  I think it was ok to go but we forgot about it. I would like to delete my branch and fix my manifest to use upstream if this could be merged.

TomzBench commented:
ok, keep your bug then ?

sambhurst commented:
@TomzBench Sorry, I closed by mistake.
